java/code/src/com/suse/utils/gson/RecordTypeAdapterFactory.java

```diff
@@ -0,0 +1,148 @@
+            return null;
+        }
+
+        return (TypeAdapter<T>) new RecordTypeAdapter<>(gson, rawType);
+    }
+
+    // Adapter for a specific record type
+    private static class RecordTypeAdapter<T> extends TypeAdapter<T> {
+
+        private final String recordName;
+
+        private final RecordComponent[] components;
+
+        private final Constructor<T> constructor;
+
+        private final Map<String, Type> componentTypeMap;
+
+        private final Gson gson;
+
+        RecordTypeAdapter(Gson gsonIn, Class<T> recordClassIn) {
+            this.gson = gsonIn;
+            this.recordName = recordClassIn.getName();
+
+            try {
+                // Find all the record components
+                this.components = recordClassIn.getRecordComponents();
+                // Cache the type of each component
+                this.componentTypeMap = Arrays.stream(components)
+                    .collect(Collectors.toMap(RecordComponent::getName, RecordComponent::getGenericType));
+                // Retrieve the canonical constructor
+                this.constructor = recordClassIn.getDeclaredConstructor(
+                    Arrays.stream(components).map(RecordComponent::getType).toArray(Class<?>[]::new)
+                );
+            }
+            catch (ReflectiveOperationException | RuntimeException ex) {
+                throw new IllegalStateException("Unable to collect the information of record " + recordName, ex);
+            }
+        }
+
+        @Override
+        public void write(JsonWriter jsonWriter, T value) throws IOException {
+            jsonWriter.beginObject();
+
+            for (RecordComponent field : components) {
+                writeField(jsonWriter, value, field);
+            }
+
+            jsonWriter.endObject();
+        }
+
+        private void writeField(JsonWriter jsonWriter, T value, RecordComponent field) throws IOException {
+            String fieldName = field.getName();
+
+            try {
+                Object fieldValue = field.getAccessor().invoke(value);
+
+                // This suppression is unavoidable as we don't know at compile-time the generic type of the component
+                @SuppressWarnings("unchecked")
+                TypeToken<Object> tokenType = (TypeToken<Object>) TypeToken.get(field.getGenericType());
+                TypeAdapter<Object> adapter = gson.getAdapter(tokenType);
+
+                jsonWriter.name(fieldName);
+                adapter.write(jsonWriter, fieldValue);
+            }
+            catch (ReflectiveOperationException | RuntimeException ex) {
+                throw new IOException("Unable to process field '%s' of record %s".formatted(fieldName, recordName), ex);
+            }
+        }
+
+        @Override
+        public T read(JsonReader jsonReader) throws IOException {
+            if (jsonReader.peek() == JsonToken.NULL) {
+                jsonReader.nextNull();
+                return null;
```

Comment:
small nitpick i think we should fail on null values to make sure if we expect a value we always get one and leave the null handling to our optional type adapter so its explicit what is required and optional.